src/compiler/types.ts

```diff
@@ -2388,6 +2388,8 @@ namespace ts {
         parameters: Symbol[];               // Parameters
         thisType?: Type;                    // type of this-type
         /* @internal */
+        thisParameter?: Symbol;             // symbol of this-type parameter

```

Comment:
It was never quite right to resolve `thisType` in getSignatureOfDeclaration` so switching to a symbol is an improvement. But it means that you should remove `thisType` entirely and replace usages with `getTypeOfSymbol(sig.thisParameter)`.